src/include/denc.h

```diff
@@ -0,0 +1,1176 @@
+                                 traits::featured>::type
+    encode(const std::set<T>& s, buffer::list::contiguous_appender& p,
+	   uint64_t f) {
+    denc((uint32_t)s.size(), p);
+    for (const T& e : s) {
+      denc(e, p, f);
+    }
+  }
+  static void decode(std::set<T>& s, buffer::ptr::iterator& p) {
+    uint32_t num;
+    denc(num, p);
+    while (num--) {
+      T temp;
+      denc(temp, p);
+      s.insert(temp);
+    }
+  }
+};
+
+//
+// std::map<A, B>
+//
+template<typename A, typename B>
+struct denc_traits<
+  std::map<A, B>,
+  typename std::enable_if<denc_traits<A>::supported &&
+			  denc_traits<B>::supported>::type> {
+  typedef denc_traits<A> a_traits;
+  typedef denc_traits<B> b_traits;
+
+  enum { supported = true };
+  enum { featured = a_traits::featured || b_traits::featured };
+  enum { bounded = a_traits::bounded && b_traits::bounded };
+
+  template<typename AA=A>
+  static typename std::enable_if<sizeof(AA) &&
+				 !bounded &&
+				 !featured>::type
+  bound_encode(const std::map<A,B>& v, size_t& p) {
+    denc((uint32_t)v.size(), p);
+    for (const auto& i : v) {
+      denc(i.first, p);
+      denc(i.second, p);
+    }
+  }
+  template<typename AA=A>
+  static typename std::enable_if<sizeof(AA) &&
+				 !bounded &&
+				 featured, void>::type
+  bound_encode(const std::map<A,B>& v, size_t& p, uint64_t f) {
+    denc((uint32_t)v.size(), p);
+    for (const auto& i : v) {
+      denc(i.first, p, f);
+      denc(i.second, p, f);
+    }
+  }
+  template<typename AA=A>
+  static typename std::enable_if<sizeof(AA) &&
+				 bounded &&
+				 !featured>::type
+  bound_encode(const std::map<A,B>& v, size_t& p) {
+    denc((uint32_t)v.size(), p);
+    size_t elem_size = 0;
+    denc(*(A*)nullptr, elem_size);
+    denc(*(B*)nullptr, elem_size);
+    p += v.size() * elem_size;
+  }
+  template<typename AA=A>
+  static typename std::enable_if<sizeof(AA) &&
+				 bounded &&
+				 featured, void>::type
+  bound_encode(const std::map<A,B>& v, size_t& p, uint64_t f) {
+    denc((uint32_t)v.size(), p);
+    size_t elem_size = 0;
+    denc(*(A*)nullptr, elem_size, f);
+    denc(*(B*)nullptr, elem_size, f);
+    p += v.size() * elem_size;
+  }
+
+  template<typename AA=A>
+  static typename std::enable_if<sizeof(AA) &&
+				 !featured>::type
+  encode(const std::map<A,B>& v, bufferlist::contiguous_appender& p) {
+    denc((uint32_t)v.size(), p);
+    for (const auto& i : v) {
+      denc(i.first, p);
+      denc(i.second, p);
+    }
+  }
+  template<typename AA=A>
+  static typename std::enable_if<sizeof(AA) &&
+				 featured, void>::type
+  encode(const std::map<A,B>& v, bufferlist::contiguous_appender& p,
+	   uint64_t f) {
+    denc((uint32_t)v.size(), p);
+    for (const auto& i : v) {
+      denc(i.first, p, f);
+      denc(i.second, p, f);
+    }
+  }
+
+  static void decode(std::map<A,B>& v, buffer::ptr::iterator& p) {
+    uint32_t num;
+    denc(num, p);
+    v.clear();
+    while (num--) {
+      A key;
```

Comment:
Should move "A key" out of the loop. If it's something like a string, then you'll likely save some mallocs.